src/bin/pg_autoctl/monitor.c

```diff
@@ -2031,6 +2115,247 @@ monitor_get_notifications(Monitor *monitor)
 }
 
 
+/*
+ * monitor_wait_until_primary_applied_settings receives notifications and
+ * watches for the following "apply_settings" set of transitions:
+ *
+ *  - primary/apply_settings
+ *  - apply_settings/apply_settings
+ *  - apply_settings/primary
+ *  - primary/primary
+ *
+ * If we lose the monitor connection while watching for the transition steps
+ * then we stop watching. It's a best effort attempt at having the CLI be
+ * useful for its user, the main one being the test suite.
+ */
+bool
+monitor_wait_until_primary_applied_settings(Monitor *monitor,
+											const char *formation)
+{
+	PGconn *connection = monitor->pgsql.connection;
+	bool		applySettingsTransitionInProgress = false;
+	bool		applySettingsTransitionDone = false;
+
+	if (connection == NULL)
+	{
+ 		log_warn("Lost connection.");
+		return false;
+	}
+
+	log_info("Waiting for the settings to have been applied to "
+			 "the monitor and primary node");
+
+	while (!applySettingsTransitionDone)
+	{
+		/* Sleep until something happens on the connection. */
+		int         sock;
+		fd_set      input_mask;
+		PGnotify   *notify;
+
+		sock = PQsocket(connection);
+
+		if (sock < 0)
+			return false;	/* shouldn't happen */
+
+		FD_ZERO(&input_mask);
+		FD_SET(sock, &input_mask);
+
+		if (select(sock + 1, &input_mask, NULL, NULL, NULL) < 0)
```

Comment:
this does not feel right. We opened the connection using libpq functions.  Can we continue with libpq ?

Reply:
I am basing this code on the Postgres documentation example code for LISTEN/NOTIFY, as seen at https://www.postgresql.org/docs/current/libpq-example.html#LIBPQ-EXAMPLE-2. I'm going to add a comment in the code that makes this obvious.